apiserver/internal/handlers/resource/resource.go

```diff
@@ -0,0 +1,408 @@
+}
+
+func extractFilename(req *http.Request) (string, error) {
+	disp := req.Header.Get(api.HeaderContentDisposition)
+
+	// the first value returned here is the media type Name (e.g. "form-data"),
+	// but we don't really care.
+	_, vals, err := mime.ParseMediaType(disp)
+	if err != nil {
+		return "", errors.Errorf("badly formatted Content-Disposition: %w", err)
+	}
+
+	param, ok := vals[api.FilenameParamForContentDispositionHeader]
+	if !ok {
+		return "", errors.Errorf("missing %q in resource upload request",
+			api.FilenameParamForContentDispositionHeader)
+	}
+
+	// Decode param, possibly encoded in base64.
+	var filename string
+	filename, err = new(mime.WordDecoder).Decode(param)
+	if err != nil {
+		// If encoding is not required, the encoder will return the original string.
+		// However, the decoder doesn't expect that, so it barfs on non-encoded
+		// strings. To detect if a string was not encoded, we simply try encoding
+		// again, if it returns the same string, we know it wasn't encoded.
+		if param == mime.BEncoding.Encode("utf-8", param) {
+			filename = param
+		} else {
+			return "", errors.Errorf("decoding filename %q from upload request: %w", param, err)
+		}
+	}
+
+	return filename, nil
+}
+
+func extractSize(req *http.Request) (int64, error) {
+	var size int64
+	if req.Header.Get(api.HeaderContentLength) == "" {
+		size = req.ContentLength
+		// size will be negative if there is no content.
+		if size < 0 {
+			size = 0
+		}
+	} else {
+		sizeRaw := req.Header.Get(api.HeaderContentLength)
+		var err error
+		size, err = strconv.ParseInt(sizeRaw, 10, 64)
+		if err != nil {
+			return 0, errors.Errorf("parsing size: %w", err)
+		}
+	}
+
+	return size, nil
+}
+
+func tagToUsername(tag names.Tag) string {
+	switch tag := tag.(type) {
+	case names.UserTag:
+		return tag.Name()
+	default:
+		return ""
+	}
+}
+
+// sendError sends a JSON-encoded error response
+// for errors encountered during processing.
+func sendError(w http.ResponseWriter, errToSend error, logger logger.Logger) error {
+	paramsErr, statusCode := apiservererrors.ServerErrorAndStatus(errToSend)
+	logger.Debugf("sending error: %d %v", statusCode, paramsErr)
+	return errors.Capture(sendStatusAndJSON(w, statusCode, &params.ErrorResult{
+		Error: paramsErr,
+	}))
+}
+
+// sendStatusAndJSON sends an HTTP status code and
+// a JSON-encoded response to a client.
+func sendStatusAndJSON(w http.ResponseWriter, statusCode int, response interface{}) error {
+	body, err := json.Marshal(response)
+	if err != nil {
+		return errors.Errorf("cannot marshal JSON result %#v: %v", response, err)
+	}
+
+	if statusCode == http.StatusUnauthorized {
+		w.Header().Set("WWW-Authenticate", `Basic realm="juju"`)
+	}
+	w.Header().Set("Content-Type", params.ContentTypeJSON)
+	w.Header().Set("Content-Length", fmt.Sprint(len(body)))
+	w.WriteHeader(statusCode)
+	if _, err := w.Write(body); err != nil {
+		return errors.Errorf("cannot write response: %w", err)
+	}
+	return nil
+}
```

Comment:
Move these to `internal/http/errors.go`.

See: https://github.com/juju/juju/pull/18727/files#diff-12d9f035db04c51ab63eba1d3317c385bd1274b5f881bc6a550e650039618ff0

I'm following your lead.